src/backend/renderer.js

```diff
@@ -1979,14 +2044,152 @@ export function attach(
+  // The return value signals whether we should keep matching siblings or not.
+  function updateTrackedPathStateBeforeMount(fiber: Fiber): boolean {
+    if (trackedPath === null || !mightBeOnTrackedPath) {
+      // Fast path: there's nothing to track so do nothing and ignore siblings.
+      return false;
+    }
+    const returnFiber = fiber.return;
+    const returnAlternate = returnFiber !== null ? returnFiber.alternate : null;
+    // By now we know there's some selection to restore, and this is a new Fiber.
+    // Is this newly mounted Fiber a direct child of the current best match?
+    // (This will also be true for new roots if we haven't matched anything yet.)
+    if (
+      trackedPathMatchFiber === returnFiber ||
+      (trackedPathMatchFiber === returnAlternate && returnAlternate !== null)
+    ) {
+      // Is this the next Fiber we should select? Let's compare the frames.
+      const actualFrame = getPathFrame(fiber);
+      const expectedFrame = trackedPath[trackedPathMatchDepth + 1];
+      if (expectedFrame === undefined) {
+        throw new Error('Expected to see a frame at the next depth.');
+      }
+      if (
+        actualFrame.index === expectedFrame.index &&
+        actualFrame.key === expectedFrame.key &&
+        actualFrame.displayName === expectedFrame.displayName
+      ) {
+        // We have our next match.
+        trackedPathMatchFiber = fiber;
+        trackedPathMatchDepth++;
+        // Are we out of frames to match?
+        if (trackedPathMatchDepth === trackedPath.length - 1) {
+          // There's nothing that can possibly match afterwards.
+          // Don't check the children.
+          mightBeOnTrackedPath = false;
+        } else {
+          // Check the children, as they might reveal the next match.
+          mightBeOnTrackedPath = true;
+        }
+        // In either case, since we have a match, we don't need
+        // to check the siblings. They'll never match.
+        return false;
+      }
+    }
+    // This Fiber's parent is on the path, but this Fiber itself isn't.
+    // There's no need to check its children--they won't be on the path either.
+    mightBeOnTrackedPath = false;
+    // However, one of its siblings may be on the path so keep searching.
+    return true;
+  }
+
+  function updateTrackedPathStateAfterMount(mightSiblingsBeOnTrackedPath) {
+    // updateTrackedPathStateBeforeMount() told us whether to match siblings.
+    // Now that we're entering siblings, let's use that information.
+    mightBeOnTrackedPath = mightSiblingsBeOnTrackedPath;
+  }
+
+  function getPathFrame(fiber: Fiber): PathFrame {
+    let { displayName, key, index } = getDataForFiber(fiber);
+    if (fiber.tag === HostRoot) {
+      // Roots don't have a real index.
+      // Instead, we'll use the order in which it mounted.
+      const id = getFiberID(getPrimaryFiber(fiber));
+      const order = rootInsertionOrder.get(id);
```

Comment:
I’m not sure I understand the suggestion. There can only be one selected root, no? Can you write up the data structure you imagine?

Treating all roots as index 0 would break any multi root app because it would always “go into” the first root if it’s root element is something generic like a div. You can try it on this fixture or on FB.

Reply:
Ah, when you  say  "go  into" I think you're referring to the fact  that  we expand/select gradually, as we find a deeper match. That makes sense.

When I  wrote this comment, I was  thinking  we  waited for the full  match before we   expanded, so I thought  we could keep  a set of possible matchers (0-1 per root) and expand on the first full path match.

It's okay though. I don't feel strongly  about  this suggestion. I was just thinking out  loud. 😄

Reply:
I see. I intentionally don't wait for the full match because in many cases just getting "close enough" is a better experience than not getting anywhere (IMO). Such as for cases like dropdowns when items don't exist after refresh.

Reply:
That makes sense.